sound/soc/sof/sof-audio.c

```diff
@@ -169,6 +169,244 @@ static int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swid
 	return ret;
 }
 
+static struct snd_sof_widget *sof_find_pipeline_by_id(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler && swidget->pipeline_id == pipeline_id)
+			return swidget;
+
+	return NULL;
+}
+
+static int sof_setup_pipeline_widget(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *pipe_widget = sof_find_pipeline_by_id(sdev, pipeline_id);
+
+	if (!pipe_widget) {
+		dev_err(sdev->dev, "error: invalid pipe widget\n");
+		return -EINVAL;
+	}
+
+	return sof_widget_setup(sdev, pipe_widget);
+}
+
+static int sof_free_pipeline_widget(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *pipe_widget = sof_find_pipeline_by_id(sdev, pipeline_id);
+
+	if (!pipe_widget)
+		return -EINVAL;
+
+	return sof_widget_free(sdev, pipe_widget);
+}
+
+static int sof_route_setup(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *wsource,
+			   struct snd_soc_dapm_widget *wsink)
+{
+	struct snd_sof_widget *src_widget = wsource->dobj.private;
+	struct snd_sof_widget *sink_widget = wsink->dobj.private;
+	struct sof_ipc_pipe_comp_connect connect;
+	struct sof_ipc_reply reply;
+	int ret;
+
+	connect.hdr.size = sizeof(connect);
+	connect.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_CONNECT;
+	connect.source_id = src_widget->comp_id;
+	connect.sink_id = sink_widget->comp_id;
+
+	/* send ipc */
+	ret = sof_ipc_tx_message(sdev->ipc, connect.hdr.cmd, &connect, sizeof(connect),
+				 &reply, sizeof(reply));
+	if (ret < 0)
+		dev_err(sdev->dev, "error: failed to load route source %s -> %s sink\n",
+			wsource->name, wsink->name);
+
+	return ret;
+}
+
+static bool is_sof_widget(struct snd_soc_dapm_widget *widget)
+{
+	switch (widget->id) {
+	case snd_soc_dapm_dai_in:
+	case snd_soc_dapm_dai_out:
+	case snd_soc_dapm_mixer:
+	case snd_soc_dapm_pga:
+	case snd_soc_dapm_buffer:
+	case snd_soc_dapm_scheduler:
+	case snd_soc_dapm_aif_out:
+	case snd_soc_dapm_aif_in:
+	case snd_soc_dapm_src:
+	case snd_soc_dapm_asrc:
+	case snd_soc_dapm_siggen:
+	case snd_soc_dapm_effect:
+	case snd_soc_dapm_mux:
+	case snd_soc_dapm_demux:
```

Comment:
This is the only way I could think of @kv2019i . Im all ears if you have a better suggestion though

Reply:
@plbossart the number of widgets SOF doesnt support is long that what it does. If I flip the logic the function is going to look a lot bigger for no good reason.

Reply:
@ranj063 Old thread, but I guess this is best. I was thinking of two approaches: 1) just let all types through, and 2) use some field like "widget->dobj.private". It would seem (1) would lead to errors, and (2) you already use, but it's no limiting enough.

I guess one option to do a look up on whether "dobj.private" is found in "sdev->widget_list" (or "w->id == snd_soc_dapm_effect" as these are not on the list). Not very nice either, but would work more seamlessly if new widget types are added to SOF.